### PR TITLE
Handle zero-length OK deflate responses (master)

### LINKED ISSUE #922: Bug: Zero length deflate responses are never fulfilled
Some older ISS servers can send zero-length deflate responses. This causes the response to ‘slip through the cracks’, leading to the response promise to never be fulfilled.

I created a PR #903 to fix this over a month ago, but I am creating an issue due to inactivity.

**Reproduction**

More detail can be be found in the PR. This affects both version 2.x and 3.x. Essentially any response similar to

```http
HTTP/1.1 200 OK
Content-Encoding: deflate
Content-Length: 0
Server: Microsoft-IIS/7.5
```

**Expected behaviour**

The node-fetch response promise should be fulfilled.

### LINKED ISSUE #1442: Fetch never resolves on empty response with `deflate`
```js
const fetch = require('node-fetch');
const http = require('http');

const server = http.createServer((request, response) => {
        response.setHeader('content-encoding', 'deflate'); // change this to gzip and it works
        response.end();
});

server.listen(8080);

(async () => {
        const response = await fetch('http://localhost:8080');
        const text = await response.text();

        console.log('Result:', text);

        server.close();
})();
```

Looks like you're missing

```diff
+				let hasData = false;
				raw.once('data', chunk => {
+					hasData = true;
					// See http://stackoverflow.com/questions/37519828
					body = (chunk[0] & 0x0F) === 0x08 ? pump(body, zlib.createInflate(), reject) : pump(body, zlib.createInflateRaw(), reject);

					response = new Response(body, responseOptions);
					resolve(response);
				});
+
+				raw.once('end', () => {
+					if (!hasData) {
+						response = new Response(body, responseOptions);
+						resolve(response);
+					}
+				});
```

before https://github.com/node-fetch/node-fetch/blob/41f53b9065a00bc73d24215d42aacdcd284b199c/src/index.js#L292

### PR DESCRIPTION
**Pull request by mgoszcz2**
**What is the purpose of this pull request?**

- [x] Bug fix

**What changes did you make?**

This fixes an issue where older ISS servers (Definitely 7.5, but I can’t rule out other versions), could respond with a zero-length OK deflate response, hanging node-fetch. This was caused by the body stream never emitting a `data` event. A typical problematic response looks something like this:

```http
HTTP/1.1 200 OK
Content-Encoding: deflate
Content-Length: 0
Server: Microsoft-IIS/7.5
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
```

**Is there anything you'd like reviewers to know?**

This is the master counterpart to the 2.x PR #903
I had to tweak it a bit compared to the 2.x version because `pump` would incorrectly reject on empty responses. 

- fix #1442
- fix #922
- fix #903

### CODE REVIEW

**Comment by mgoszcz2** [comment]
Not really sure how to fix the coverage here. In the original PR adding `pump` (#670), @xxczaki [mentions](https://github.com/node-fetch/node-fetch/pull/670#issuecomment-529225954) that the stream error conditions are hard to test.

**Comment by uday79936** [comment]
HTTP/1.1 200 OK
Content-Encoding: deflate
Content-Length: 0
Server: Microsoft-IIS/7.5
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET is change first and merge is accept

**Comment by mgoszcz2** [comment]
Don't see what you're saying here, looks like it might be Hacktoberfest spam?

**Comment by mgoszcz2** [comment]
Ping @jimmywarting as it’s been a couple weeks. This addresses your comment https://github.com/node-fetch/node-fetch/pull/903#issuecomment-690946225

File: src/index.js
Comment (bitinn) [inline src/index.js:188]: Sorry for the wait, this PR looks fine to me, but I think a few things can be improved:
 
- If the change on `pump()` error handling is necessary for your tests to pass, then we should probably change all of the instances.

File: src/index.js
Comment (bitinn) [inline src/index.js:265]: - I would like to see a comment with link to explain special handling of deflate response, so we have a source for the code, see my comment above that explains the line `if ((chunk[0] & 0x0F) === 0x08)`

**Review by bitinn — CHANGES_REQUESTED** [review]
Sorry for the wait, this PR looks fine to me, but I think a few things can be improved.
 
- If the change on `pump()` error handling is necessary for your tests to pass, then we should probably change all of the instances.

- I would like to see a comment with link to explain special handling of deflate response, so we have a source of the issues

**Comment by xxczaki** [comment]
@mgoszcz2 Is it ready for review now?

**Comment by mgoszcz2** [comment]
I think so @xxczaki. No bother at all, you can see I was in no rush either.

**Comment by mgoszcz2** [comment]
Hi, any chance of this getting merged? @xxczaki @bitinn

**Comment by tekwiz** [comment]
@mgoszcz2 This just needs to be rebased on main.  I can approve once that's done, and I believe @bitinn just needs to re-review after that and we can merge.

**Comment by mgoszcz2** [comment]
@tekwiz Thanks. I hope this also means #903 will be merged? Because there was a lot of churn involved in getting to this stage without any merges happening and I don't feel like sinking more time into this pull request. I made #903 to fix a specific issues that was affecting production at a company I no longer work. Unless nothing more is required of me other than a rebase, I'm not willing to continue working on this PR.

**Comment by bitinn** [comment]
I believe this is addressing the same issue as

https://github.com/node-fetch/node-fetch/issues/1442

I hope Jeff can help to look into it. We are now 2 years away from my initial review, I am happy to merge but we need to make sure it doesn’t introduce any problems to current codebase.

**Review by bitinn — APPROVED** [review]

**Comment by mgoszcz2** [comment]
Can you explain what's preventing from #903 from being merged? It's a completely different branch and major version. I wasn't even using 3.x, I only made this PR because I was asked to by Jimmy.

**Comment by mgoszcz2** [comment]
I rebased it

**Comment by mgoszcz2** [comment]
Regarding, #1442: Yes it's the same issue, and in fact that it's a duplicate of #922 I opened in 2020. In case it's not clear, this PR would've fixed it on the main branch, and #903 would've fixed it on the 2.x branch (but nether has been merged).

**Comment by bitinn** [comment]
Thx, for 2.x branch backport we can reopen your PR, please ping Jimmy in the comment. I think a lack of active reviewers at the time was preventing it from moving forward.

**Review by jimmywarting — APPROVED** [review]

File: src/index.js
Comment (jimmywarting) [inline src/index.js:289]: nit pick, can you quickly change all `err` to `error` to make the linter happy?

must say that i'm not a huge fan of xo... like standardjs more...

**Review by JefferyHus — APPROVED** [review]
Looks good to me. With only changing `err` to `error`

**Comment by mgoszcz2** [comment]
Done. Please also consider merging/re-opening #903 when you have some time.

**Comment by jimmywarting** [comment]
@mgoszcz2

> Done. Please also consider merging/re-opening #903 when you have some time.

haven't looked trough (reviewed) it yet... it's closed... what dose that pr do that this pr dose not solve?

**Comment by mgoszcz2** [comment]
It fixes this bug for the 2.x branch. You closed it. I do not understand why. When I asked if it will be merged and you said it will be. That was September 2020. I only made this PR because you asked for it to be ported to the main branch too. Please look over your own messages in #903. In retrospect I should have never offered to fix the main branch if I knew merging 6 lines of code in two branches would take nearly two years.

**Comment by mgoszcz2** [comment]
I pinged you at the request of David. Please read the messages above.

> Thx, for 2.x branch backport we can reopen your PR, please ping Jimmy in the comment. I think a lack of active reviewers at the time was preventing it from moving forward.

**Comment by jimmywarting** [comment]
merged, thx for the fix, again sry that it took so long

**Pull request merged**
